zerver/tests/test_thumbnail.py

```diff
@@ -0,0 +1,207 @@
+
+        # Test with another user trying to access image using thumbor.
+        self.logout()
+        self.login(self.example_email("iago"))
+        result = self.client_get("/thumbnail" + uri)
+        self.assertEqual(result.status_code, 403, result)
+        self.assert_in_response("You are not authorized to view this file.", result)
+
+    @override_settings(THUMBOR_HOST='127.0.0.1:9995')
+    def test_with_static_files(self):
+        # type: () -> None
+        self.login(self.example_email("hamlet"))
+        uri = '/static/images/cute/turtle.png'
+        result = self.client_get("/thumbnail" + uri)
+        self.assertEqual(result.status_code, 302, result)
+        self.assertEqual(uri, result.url)
+
+    def test_with_thumbor_disabled(self):
+        # type: () -> None
+        self.login(self.example_email("hamlet"))
+        fp = StringIO("zulip!")
+        fp.name = "zulip.jpeg"
+
+        result = self.client_post("/json/user_uploads", {'file': fp})
+        self.assert_json_success(result)
+        json = ujson.loads(result.content)
+        self.assertIn("uri", json)
+        uri = json["uri"]
+        base = '/user_uploads/'
+        self.assertEqual(base, uri[:len(base)])
+
+        with self.settings(THUMBOR_HOST=''):
+            result = self.client_get("/thumbnail" + uri)
+        self.assertEqual(result.status_code, 302, result)
+        base = '/user_uploads/'
+        self.assertEqual(base, result.url[:len(base)])
+
+        uri = 'https://www.google.com/images/srpr/logo4w.png'
+        with self.settings(THUMBOR_HOST=''):
+            result = self.client_get("/thumbnail/" + uri)
+        self.assertEqual(result.status_code, 302, result)
+        self.assertEqual(uri, result.url)
+
+        uri = 'http://www.google.com/images/srpr/logo4w.png'
+        with self.settings(THUMBOR_HOST=''):
+            result = self.client_get("/thumbnail/" + uri)
+        self.assertEqual(result.status_code, 302, result)
+        base = 'https://external-content.zulipcdn.net/7b6552b60c635e41e8f6daeb36d88afc4eabde79/687474703a2f2f7777772e676f6f676c652e636f6d2f696d616765732f737270722f6c6f676f34772e706e67'
+        self.assertEqual(base, result.url)
+
+    def test_with_different_thumbor_host(self):
+        # type: () -> None
+        self.login(self.example_email("hamlet"))
+        fp = StringIO("zulip!")
+        fp.name = "zulip.jpeg"
+
+        result = self.client_post("/json/user_uploads", {'file': fp})
+        self.assert_json_success(result)
+        json = ujson.loads(result.content)
+        self.assertIn("uri", json)
+        uri = json["uri"]
+        base = '/user_uploads/'
+        self.assertEqual(base, uri[:len(base)])
+
+        with self.settings(THUMBOR_HOST='test-thumborhost.com'):
+            result = self.client_get("/thumbnail" + uri)
+        self.assertEqual(result.status_code, 302, result)
+        base = 'test-thumborhost.com/thumbor/unsafe/0x0/smart/files/'
```

Comment:
why not verify the entire URL here?

Reply:
Since the entire url will contain a dynamic sign key it is kind of tricky to do the testing without generating the entire hash again here using a copy of logic earlier used.